### PR TITLE
add api definition

### PR DESCRIPTION
**Pull request by Deardrops**
Add api definitions:

- Repo
- BackupPolicy
- BackupJob
- RestoreJob
- BackupGroup
- Backup

### CODE REVIEW

File: hack/boilerplate.go.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/*
+Copyright 2021.
```
Comment (kylezh) [inline hack/boilerplate.go.txt:2]: Copyright 2021 The IOMesh Authors.

File: PROJECT [hunk 2]
```diff
@@ -0,0 +1,70 @@
+domain: smartx.com
```
Comment (kylezh) [inline PROJECT:1]: domain: iomesh.com

File: PROJECT [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+layout:
+- go.kubebuilder.io/v3
+projectName: backuptool
```
Comment (kylezh) [inline PROJECT:4]: projectName: backup

File: PROJECT [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+repo: smartx.com/backuptool
```
Comment (kylezh) [inline PROJECT:5]: iomesh.com/backup

File: PROJECT [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+resources:
+- api:
+    crdVersion: v1
+    namespaced: true
+  controller: true
+  domain: smartx.com
```
Comment (kylezh) [inline PROJECT:11]: iomesh.com

File: PROJECT [hunk 6]
```diff
@@ -1,70 +1,70 @@
...
+domain: iomesh.com
 layout:
 - go.kubebuilder.io/v3
-projectName: backuptool
-repo: smartx.com/backuptool
+projectName: backup
+repo: iomesh.com/backup
 resources:
 - api:
     crdVersion: v1
```
Comment (kylezh) [inline PROJECT:8]: crdVersion 应该都是 v1alpha1

File: api/v1alpha1/backupgroup_types.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+type BackupGroupSpec struct {
+	Source      Target `json:"source,omitempty"`
+	Repo        string `json:"repo,omitempty"` // Repo name
+	GlobalDedup bool   `json:"globalDedup,omitempty"`
+}
+
+// BackupGroupStatus defines the observed state of BackupGroup
+type BackupGroupStatus struct {
+	RunningBackups  []BackupID `json:"runningBackups,omitempty"`
+	CompetedBackups []BackupID `json:"competedBackups,omitempty"`
```
Comment (kylezh) [inline api/v1alpha1/backupgroup_types.go:33]: `completedBackups`

File: api/v1alpha1/common.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+type Target string
+
+type JobID uint64
+
+type BackupGroupID uint64
+
+type BackupID uint64
+
+// For dedup store only
+type SnapshotID uint64
```
Comment (kylezh) [inline api/v1alpha1/common.go:32]: 所有的 ID 类型都应该是 string，实际内容是随机生成的 UUID。

使用的时候，用 Name 来表示，例如：JobName、BackupGroupName

**Comment by kylezh** [comment]
需要补充 github action

File: PROJECT (hunk 6 above)
Comment (Deardrops) [reply PROJECT:8]: `resources.api.crdVersion` means

> The Kubernetes API version (apiVersion) used to do the scaffolding for the CRD resource.

Ref: https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/reference/project-config.md

File: api/v1alpha1/backup_types.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BackupSpec defines the desired state of Backup
+type BackupSpec struct {
+	Kind             BackupType    `json:"kind,omitempty"`
+	BackupGroupName  BackupGroupID `json:"backupGroupName,omitempty"`
+	ParentBackupName BackupID      `json:"parentBackupName,omitempty"`
+	BackupJobName    JobID         `json:"backupJobName,omitempty"`
```
Comment (kylezh) [inline api/v1alpha1/backup_types.go:28]: It is wired to use `XXXID` as type name while `XXXName` as variable name.

For most cases, it is not necessary to have a new type named `XXXID` for referencing another object, just `string` is ok. If one `string` filed is not enough, there would be a struct like `XXXRef` as a reference type:

```
type XXXRef struct {
  Name string
  XXX string
}
```

File: api/v1alpha1/common.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+type JobID = UUID
+
+type BackupGroupID = UUID
+
+type BackupID = UUID
+
+// For dedup store only
+type SnapshotID = UUID
+
```
Comment (kylezh) [inline api/v1alpha1/common.go:35]: As comment at another place, these type definitions are not necessary.

File: api/v1alpha1/common.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+type JobMeta struct {
+	ScheduledTime metav1.Time `json:"scheduledTime,omitempty"`
+	Pause         bool        `json:"pause,omitempty"`
```
Comment (kylezh) [inline api/v1alpha1/common.go:38]: `Pause` -> `Paused`

File: api/v1alpha1/restorejob_types.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// RestoreJobSpec defines the desired state of RestoreJob
+type RestoreJobSpec struct {
+	JobMeta `json:"jobMeta,omitempty"`
+
+	From Target `json:"from,omitempty"`
```
Comment (kylezh) [inline api/v1alpha1/restorejob_types.go:27]: `From` should be `BackupName`?

File: api/v1alpha1/backupjob_types.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BackupJobSpec defines the desired state of BackupJob
+type BackupJobSpec struct {
+	JobMeta `json:"jobMeta,omitempty"`
+
+	From Target `json:"from,omitempty"`
+	To   Target `json:"to,omitempty"`
```
Comment (kylezh) [inline api/v1alpha1/backupjob_types.go:28]: `To` should be `BackupName`?

Comment (Deardrops) [reply api/v1alpha1/backupjob_types.go:28]: No, `To` should be repo name + backup name, such as `repo://<repo-name>/backup`
Here we should keep use the `Target`(string) type.

File: api/v1alpha1/restorejob_types.go (hunk 12 above)
Comment (Deardrops) [reply api/v1alpha1/restorejob_types.go:27]: same as `BackupJobSpec.To`

File: api/v1alpha1/backupjob_types.go (hunk 13 above)
Comment (kylezh) [reply api/v1alpha1/backupjob_types.go:28]: I think `BackupName` is enough. We can fetch `Repo` info from `Backup` -> `BackupGroup` -> `Repo`.

Comment (Deardrops) [reply api/v1alpha1/backupjob_types.go:28]: Sure, I will change it to `BackupName`.

**Comment by Deardrops** [comment]
> 
> 
> 需要补充 github action

@kylezh github action have been added.

Now we support `make test` and `make docker-build`.

File: api/v1alpha1/backupjob_types.go (hunk 13 above)
Comment (kylezh) [reply api/v1alpha1/backupjob_types.go:28]: `BackupGroup` should be created before any `BackupJob`. User would select a `BackupGroup`, and then create a `BackupJob` for backing up its data.

Comment (Deardrops) [reply api/v1alpha1/backupjob_types.go:28]: Ok, we reach an agreement with:
```yaml
backupJob:
	Spec:
		BackupGroupName: xxxx
restoreJob:
	Spec:
		from: xxxxx # BackupName
		to: zbs://<hosts>/<volume-id>
```

**Comment by kylezh** [comment]
LGTM. Please squash unnecessary commits.

**Pull request merged**